proposals/0066-feature-gate-threshold-automation.md

```diff
@@ -0,0 +1,245 @@
+
+## Detailed Design
+
+Before this SIMD is published, we will deploy a BPF program to manage feature
+activations, which will allow us to revoke pending feature activations
+(previously not possible).
+The feature activation process will otherwise be unchanged, so no SIMD is
+required.
+This program will be upgradeable; the authority and process through which it
+will be upgraded is outside the scope of this SIMD. Since this SIMD requires an
+update to the program, the upgrade details will be determined in parallel.
+
+The initial Feature Gate Program can be found in this pull request to the
+Solana Program Library:
+<https://github.com/solana-labs/solana-program-library/pull/4987>.
+The runtime changes and feature activation that will enable this program can be
+found in this feature tracking issue:
+<https://github.com/solana-labs/solana/issues/33547>.
+
+### Overview
+
+Consider two arbitrary versions of the Solana runtime, where the newer version
+has more features in its feature set.
+
+| v1.14.25  | v1.16.0   |
+| --------- | --------- |
+| Feature A | Feature A |
+|           | Feature B |
+| Feature C | Feature C |
+|           | Feature D |
+
+The runtime should activate each of these features when a strong majority of
+the network supports it. “Support” for a feature means that a node is running a
+version of the Solana runtime where the code required for the feature is
+present. For instance, on a cluster with 50% of stake running v1.14.25 and 50%
+of stake running v1.16.0, only Features A and C should be eligible for
+activation.
+
+This SIMD proposes 3 steps over the span of one epoch to determine which
+feature gates are queued for activation and assess stake support:
+
+1. On an epoch boundary, the runtime generates a list of feature-gates queued
+for activation.
+2. Validators signal which of the queued feature-gates they support.
+3. On the next epoch boundary, the runtime activates the feature-gates that
+have the necessary stake support
+
+### Representing the Current Feature Set
+
+As mentioned above, a BPF program is being deployed that will replace the
+non-existent account `Feature111111111111111111111111111111111111`, which is
+the owner of all feature accounts.
+
+The Feature Gate program will utilize two PDAs to track the authoritative
+feature set queued for activation. One PDA (the Feature-Request PDA) will
+store newly requested feature activations and the other (the Feature-Queue
+PDA) will store the previous epoch’s requested activations.
+
+When a new epoch begins, the Feature-Request PDA will be an empty list. When
+key-holders submit the CLI command to activate a feature, the Feature Gate
+Program appends the feature ID to the Feature-Request set at the same time as
+creating the feature account . Revoked features are removed from the
+Feature-Request list. At the end of the epoch, the Feature-Request set is
+written to the feature queue PDA, where it becomes immutable. The
```

Comment:
"written to the feature queue PDA" by whom? The runtime I presume? We should call that out

Reply:
Yep, the runtime. Added this for clarity!